nengo/spa/module.py

```diff
@@ -25,3 +51,159 @@ def on_add(self, spa):
         modules that connect to other things in the SPA model (such as
         the basal ganglia or thalamus).
         """
+
+    def __setattr__(self, key, value):
+        """A setattr that handles Modules being added specially.
+
+        This is so that we can use the variable name for the Module as
+        the name that all of the SPA system will use to access that module.
+        """
+        if hasattr(self, key) and isinstance(getattr(self, key), Module):
+            raise SpaModuleError("Cannot re-assign module-attribute %s to %s. "
+                                 "SPA module-attributes can only be assigned "
+                                 "once." % (key, value))
+        super(Module, self).__setattr__(key, value)
+        if isinstance(value, Module):
+            if value.label is None:
+                value.label = key
+            self._modules[key] = value
+            for k, (obj, v) in iteritems(value.inputs):
+                if isinstance(v, int):
+                    value.inputs[k] = (obj, self.vocabs.get_or_create(v))
+            for k, (obj, v) in iteritems(value.outputs):
+                if isinstance(v, int):
+                    value.outputs[k] = (obj, self.vocabs.get_or_create(v))
+
+            value.on_add(self)
+
+    def __exit__(self, ex_type, ex_value, traceback):
+        super(Module, self).__exit__(ex_type, ex_value, traceback)
+        if ex_type is not None:
+            # re-raise the exception that triggered this __exit__
+            return False
+
+        module_list = frozenset(self._modules.values())
+        for net in self.networks:
+            # Since there are no attributes to distinguish what's been added
+            # and what hasn't, we have to ask the network
+            if isinstance(net, Module) and (net not in module_list):
+                raise SpaModuleError("%s must be set as an attribute of "
+                                     "a SPA network" % (net))
+
+    def get_module(self, name, strip_output=False):
```

Comment:
> Yes, ambiguity is bad. I want to be explicit.

Ah, I finally figured out why I was confused about this ambiguity.  In the past, the purpose of this function was primarily to get the module *for a particular input or output*.  So I was interpretting it as the unambiguous "get me the module for whatever name I give you -- if I give you a module name, then give me the module, and if I give you an input or output inside that module, then still give me the module".  

In any case, that change seems reasonable to me, and I can't find any place that it breaks, so it's all good.  :)